M2/Macaulay2/m2/help.m2

```diff
@@ -0,0 +1,107 @@
+examples Thing     := x -> (
+    checkLoadDocumentation();
+    d := fetchRawDocumentation makeDocumentTag(x, Package => null);
+    if d =!= null and d.?Description then (stack deepSplice getExampleInputs d.Description)^-1)
+
+-----------------------------------------------------------------------------
+-- get a list of commands whose name matches the regex
+-----------------------------------------------------------------------------
+apropos = method()
+apropos String := (pattern) -> (
+    last \ sort unique select(flatten \\ pairs \ dictionaryPath,
+        (nam,sym) -> match(pattern,nam) and not match("\\$",nam)))
+
+-----------------------------------------------------------------------------
+-- get a list of commands whose documentation matches the regex
+-----------------------------------------------------------------------------
+-- TODO: can this be simplified?x
+lastabout := null
+
+matchfun := (re, db) -> key -> (
+    -- not quite right, because the body might assert that the key is undocumented.
+    -- We need a quicker way to identify undocumented keys.
+    if db === null then match(re, key)
+    -- not quite right, because this string might occur in the raw documentation as
+    -- part of the description. Unlikely, though.
+    else if instance(db, Database) then (
+        (match(re, key) or match(re, db#key)) and not match(///"undocumented" => true///, db#key))
+    else if instance(db, HashTable) then (
+        not db#key#?"undocumented" and (match(re, key) or db#key.?Description and match(re,toString db#key.Description))))
+
+about = method(Options => {Body => false})
+about Type     :=
+about Symbol   :=
+about Function := o -> f -> about("\\b" | toString f | "\\b", o)
+about String   := o -> re -> lastabout = (
+    packagesSeen := new MutableHashTable;
+    NumberedVerticalList sort join(
+        flatten for pkg in loadedPackages list (
+            pkgname := pkg#"pkgname";
+            if packagesSeen#?pkgname then continue else packagesSeen#pkgname = 1;
+            keyList := join (
+                if not pkg#?"raw documentation database" then {}
+                else select(keys pkg#"raw documentation database",
+                    matchfun_re if o.Body then pkg#"raw documentation database"),
+                select(keys pkg#"raw documentation",
+                    matchfun_re if o.Body then pkg#"raw documentation"));
+            apply(keyList, key -> pkgname | "::" | key)),
+        flatten for pkg in getPackageInfoList() list (
+            pkgname := pkg#"name";
+            if packagesSeen#?pkgname then continue else packagesSeen#pkgname = 1;
+            dbname := pkg#"doc db file name";
+            dbkeys := keys fetchDocKeys pkg;
+            db := if o.Body then openDatabase dbname;
+            keyList := select(dbkeys, matchfun_re db);
+            if o.Body then close db;
+            apply(keyList, key -> pkgname | "::" | key))))
+
+-- TODO: why doesn't help ZZ work?
+help#0 ZZ := i -> (
```

Comment:
It does work:
```m2

i2 : about resolution

o2 = {0 => ChainComplexExtras::resolution(ChainComplex)               }
     {1 => Macaulay2Doc::resolution                                   }
     {2 => Macaulay2Doc::resolution(..., DegreeLimit => ...)          }
     {3 => Macaulay2Doc::resolution(..., FastNonminimal => ...)       }
     {4 => Macaulay2Doc::resolution(..., HardDegreeLimit => ...)      }
     {5 => Macaulay2Doc::resolution(..., LengthLimit => ...)          }
     {6 => Macaulay2Doc::resolution(..., PairLimit => ...)            }
     {7 => Macaulay2Doc::resolution(..., SortStrategy => ...)         }
     {8 => Macaulay2Doc::resolution(..., StopBeforeComputation => ...)}
     {9 => Macaulay2Doc::resolution(..., Strategy => ...)             }
     {10 => Macaulay2Doc::resolution(..., SyzygyLimit => ...)         }
     {11 => Macaulay2Doc::resolution(Ideal)                           }
     {12 => Macaulay2Doc::resolution(Matrix)                          }
     {13 => Macaulay2Doc::resolution(Module)                          }
     {14 => Macaulay2Doc::resolution(MonomialIdeal)                   }
     {15 => NCAlgebra::resolution(NCMatrix)                           }

o2 : NumberedVerticalList

i3 : help 11

o3 = Macaulay2Doc::resolution(Ideal) -- compute a projective resolution of (the quotient ring corresponding to) an ideal
     *******************************************************************************************************************

     +-----------------------------------------+
     |i1 : R = ZZ[a..d]                        |
     |                                         |
     |o1 = R                                   |
     |                                         |
     |o1 : PolynomialRing                      |
     +-----------------------------------------+
     |i2 : I = ideal(a,b,c,d)                  |
     |                                         |
     |o2 = ideal (a, b, c, d)                  |
     |                                         |
     |o2 : Ideal of R                          |
     +-----------------------------------------+
     |i3 : C = res I                           |
     |                                         |
     |      1      4      6      4      1      |
     |o3 = R  <-- R  <-- R  <-- R  <-- R  <-- 0|
     |                                         |
     |     0      1      2      3      4      5|
     |                                         |
     |o3 : ChainComplex                        |
     +-----------------------------------------+
     |i4 : C_2                                 |
     |                                         |
     |      6                                  |
     |o4 = R                                   |
     |                                         |
     |o4 : R-module, free, degrees {6:2}       |
     +-----------------------------------------+
     |i5 : C.dd_2                              |
     |                                         |
     |o5 = {1} | -b 0  -c 0  0  -d |           |
     |     {1} | a  -c 0  0  -d 0  |           |
     |     {1} | 0  b  a  -d 0  0  |           |
     |     {1} | 0  0  0  c  b  a  |           |
     |                                         |
     |             4       6                   |
     |o5 : Matrix R  <--- R                    |
     +-----------------------------------------+

     See also
     ========

       * "ChainComplex _ ZZ" -- component
       * "dd" -- differential in a chain complex
       * "resolution" -- projective resolution
       * "ideal" -- make an ideal

o3 : DIV
```

Reply:
I meant why doesn't `help ZZ := i -> ...` work?

Reply:
Oh, because `help` is actually a "Command", not a method function.